### PR TITLE
(maint) prep for 0.12.0 release

### PR DESCRIPTION
**Pull request by underscorgan**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,21 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
...
 ## [Unreleased]
 
+## [0.12.0] - released on 2017-05-31
+### Added
+- Added `generate_source_artifacts` method to the platform dsl to allow generation
+  of `.src.rpm` and debian source artifacts.
+- Added the ability to specify mirror URLs.
+
+### Deprecated
+- The existing rewrite engine is being deprecated in favor of the mirror URLs.
```
Comment (mckern) [inline CHANGELOG.md]: I'd probably expand this out a little bit more; it's a meaningful change for users.

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,21 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
...
+
+### Changed
+- Updated to sending notification messages on STDERR instead of STDOUT for easier filtering.
```
Comment (mckern) [inline CHANGELOG.md]: This is also pretty meaningful so I'd probably add a couple of lines about what/why and reference the PR where it went in for context?

File: CHANGELOG.md (hunk 1 above)
Comment (underscorgan) [inline CHANGELOG.md]: Good point, will update.

File: CHANGELOG.md (hunk 2 above)
Comment (underscorgan) [inline CHANGELOG.md]: #same.

**Comment by underscorgan** [comment]
@mckern this better?

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,21 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
 
 ## [Unreleased]
 
+## [0.12.0] - released on 2017-05-31
+### Added
+- Added `generate_source_artifacts` method to the platform dsl to allow generation
+  of `.src.rpm` and debian source artifacts.
+- Added the ability to specify mirror URLs.
```
Comment (mckern) [inline CHANGELOG.md]: I'd probably change where the bulk of the detail lives. I cribbed from your Deprecated writeup and moved it into Added:

```markdown
- Added support to the Component DSL for Mirror URLs using the `#mirror` method. This functionality supersedes the previous Rewrite Rule functionality and the Rewrite Rule functionality has been deprecated (see *Deprecated* for `0.12.0`).

  Vanagon will attempt to use any Mirror URLs provided for a component before it attempts to use the canonical upstream URL. Multiple Mirror URLs can be specified, and Vanagon will attempt to retrieve them in random order until a URL returns a `200` code or all Mirror URLs are exhausted.
```

Comment (underscorgan) [inline CHANGELOG.md]: 👍  thanks!

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,21 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
...
+
+### Deprecated
+- The existing rewrite engine is being deprecated in favor of the mirror URLs. Instead of specifying rewrite rules for retrieving your source you can now specify multiple mirrors that will be tried in a random order until the source is found. For more examples of mirrors in use, see the [example project](examples/projects/project.rb) and [component](examples/components/component2.rb).
```
Comment (mckern) [inline CHANGELOG.md]:
```markdown
- The Component DSL Rewrite Engine is deprecated in favor of Component DSL Mirror URLs. Existing Rewrite Rules will be converted to Mirror URLs and added to the Mirror URL list. The Component DSL Rewrite Engine will be removed before Vanagon 1.0.0 is released.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,21 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
...
+
+### Changed
+- Updated to sending notification messages on STDERR instead of STDOUT. This both enables easier filtering, and helps with some of the fragility around commands expecting formatted output on STDOUT. For more context around this, see the [ticket](https://tickets.puppetlabs.com/browse/VANAGON-57) and [pull request](https://github.com/puppetlabs/vanagon/pull/474).
+
+### Fixes
+- Fixed backtraces from exceptions thrown in config files.
```
Comment (mckern) [inline CHANGELOG.md]:
```markdown
Exceptions now provide accurate backtraces if thrown while parsing Component, Platform, or Projects.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,23 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
...
+  Vanagon will attempt to use any Mirror URLs provided for a component before it attempts to use the canonical upstream URL. Multiple Mirror URLs can be specified, and Vanagon will attempt to retrieve them in random order until a URL returns a `200` code or all Mirror URLs are exhausted.
+  For examples of mirrors in use, see the [example project](examples/projects/project.rb) and [component](examples/components/component2.rb).
+
+### Deprecated
+- The Component DSL Rewrite Engine is deprecated in favor of Component DSL Mirror URLs. Existing Rewrite Rules will be converted to Mirror URLs and added to the Mirror URL list. The Component DSL Rewrite Engine will be removed before Vanagon 1.0.0 is released.
+
+### Changed
+- Updated to sending notification messages on STDERR instead of STDOUT. This both enables easier filtering, and helps with some of the fragility around commands expecting formatted output on STDOUT. For more context around this, see the [ticket](https://tickets.puppetlabs.com/browse/VANAGON-57) and [pull request](https://github.com/puppetlabs/vanagon/pull/474).
```
Comment (mckern) [inline CHANGELOG.md]: NITPICKING but you could lose the commas in this section. Not important though; just streamlining.

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,23 @@ This changelog adheres to [Keep a CHANGELOG](http://keepachangelog.com/).
 
 ## [Unreleased]
 
+## [0.12.0] - released on 2017-05-31
+### Added
+- Added `generate_source_artifacts` method to the platform dsl to allow generation
+  of `.src.rpm` and debian source artifacts.
+- Added support to the Component DSL for Mirror URLs using the `#mirror` method. This functionality supersedes the previous Rewrite Rule functionality and the Rewrite Rule functionality has been deprecated (see *Deprecated* for `0.12.0`).
+  Vanagon will attempt to use any Mirror URLs provided for a component before it attempts to use the canonical upstream URL. Multiple Mirror URLs can be specified, and Vanagon will attempt to retrieve them in random order until a URL returns a `200` code or all Mirror URLs are exhausted.
```
Comment (stahnma) [inline CHANGELOG.md]: Does this follow redirects? (e.g. 301/302s?)

Comment (mckern) [inline CHANGELOG.md]: Yes. All of the HTTP/HTTPS logic in Vanagon should.

**Pull request merged**